lib/holidays/use_case/context/between.rb

```diff
@@ -13,102 +13,16 @@ def call(start_date, end_date, dates_driver, regions, observed, informal)
-                      valid_range_year = true if year_range[year_range.keys.first].cover?(year)
-                    end
-                    break if valid_range_year
-                  end
-                  next unless valid_range_year
-                end
-
-                #FIXME I don't like this entire if/else. If it's a function, do something, else do some
-                # weird mday logic? Bollocks. I think this should be a refactor target.
-                if h[:function]
-                  function_arguments = []
-
-                  #FIXME This is a refactor target. We should also allow 'date'. Right now these are the only
-                  # three things that we allow in. I think having a more testable, robust approach here is vital.
-                  if h[:function_arguments].include?(:year)
-                    function_arguments << year
-                  end
-
-                  if h[:function_arguments].include?(:month)
-                    function_arguments << month
-                  end
-
-                  if h[:function_arguments].include?(:day)
-                    function_arguments << h[:mday]
-                  end
-
-                  result = call_proc(h[:function], *function_arguments)
-
-                  #FIXME This is a dangerous assumption. We should raise an error or something
-                  #      if these procs return something unexpected.
-                  #
-                  # Procs may return either Date or an integer representing mday
-                  if result.kind_of?(Date)
-                    if h[:function_modifier]
-                      result = result + h[:function_modifier] # NOTE: This could be a positive OR negative number.
-                    end
-
-                    month = result.month
-                    mday = result.mday
-                  else
-                    mday = result
-                  end
-                else
-                  mday = h[:mday] || day_of_month_calculator.call(year, month, h[:week], h[:wday])
-                end
-
-                # Silently skip bad mdays
-                begin
-                  date = Date.civil(year, month, mday)
-                rescue; next; end
-
-                #FIXME We should be checking the function arguments and passing in what is specified.
-                # Right now all 'observed' functions require 'date' but that is by convention. Nothing
-                # is requiring that. We should be more explicit.
-                if observed && h[:observed]
-                  date = call_proc(h[:observed], date)
-                end
-
-                if date.between?(start_date, end_date)
-                  holidays << {:date => date, :name => h[:name], :regions => h[:regions]}
-                end
-              end
-            end
-          end
-
+          holidays = holidays.select{|holiday|holiday[:date].between?(start_date, end_date)}
```

Comment:
Of course! I'm just happy to even have something to review. 😄 

To continue this discussion I will write out how I think about the future of this class. You (or anyone else) can ignore it if you aren't interested. It is still helpful for me to write it down.

I tend to be very 'dependency injection' heavy. This means that I like to send a lot of separate objects into the initializer and then call them in the code. I prefer that pattern  over sharing code via inheritance or via a module. So my vision for `Between` was to 

* name it better, since 'Between' isn't a great or descriptive name
* split out all of the various pieces  (like function handling or your 'year range' logic) into separate classes. 

That way those specific classes can be tested individually and the tests for `Between` really just test that we call the different components correctly rather than having one test file that tests ALL of the `Between` logic.

Here is an example:

```ruby
        def initialize(holidays_by_month_repo, day_of_month_calculator, custom_methods_repo, proc_result_cache_repo, year_range_calculator, function_modifier_handler, observed_handler, informal_handler)
          @holidays_by_month_repo = holidays_by_month_repo
          @day_of_month_calculator = day_of_month_calculator
          @custom_methods_repo = custom_methods_repo
          @proc_result_cache_repo = proc_result_cache_repo
          @year_range_calculator = year_range_calculator
          @function_modifier_handler = function_modifer_handler #this is an awful name but you get the idea
          @observed_handler = observed_handler
          @informal_handelr = informal_handler
        end
```

The big downside to this approach is that the initializer method gets to be very big. I've of the mind that as long as it is clear what each thing does it is okay but I can understand that some people wouldn't like that.

All this aside, I am a big, big believer in test-driven development and in taking small steps. We can always later. Our test coverage is getting very, very good. That's why I fully support this change. It is a definite improvement and it doesn't block us from future refactors.